### DISCUSSION TITLE
How to use puppeteer in an API route

### DISCUSSION THREAD

**Discussion by mfix22** [vercel/next.js#29856] [Help] [answered]
I am trying to migrate a Vercel node.js lambda to a Next.js API route (still deployed to Vercel). 
However, because Next.js API routes include some built-in functionality, I am now over the 50mb size maximum, despite only using `chrome-aws-lambda` and `puppeteer-core`: https://github.com/carbon-app/carbon/pull/1262

Anyone have any suggestions for slimming down this build further?

**Comment by samcx** [comment, marked as answer]
Hi there,

We don't recommend using `chrome-aws-lambda` and `puppeteer-core` in a Serverless Function since you are already veering very close to, if not over, 50 MB.

- https://packagephobia.com/result?p=chrome-aws-lambda
- https://packagephobia.com/result?p=puppeteer-core

**Reply by mfix22** [reply]
@samsisle then how would you suggest I use it with Next.js? It works perfectly fine in a regular Vercel lambda (as you can see by the PR), but when I use it in a Next.js API route, it goes over the limit. Is there a way to disable including the Next.js bundle in the lambda?

**Reply by samcx** [reply]
It's totally valid to use a separate `/api` folder at the root of your project directory (or in a separate project) — Vercel will create those separate to your Next.js API routes.

**Reply by mfix22** [reply]
Is there a way to do that and get automatic params though? E.g. `/pages/api/images/[id].js` won't work when I move it to `/api/images/[id].js` without setting up a custom `routes` config in `vercel.json`, right?

**Reply by samcx** [reply]
You can do this without needing a `vercel.json` → https://vercel.com/docs/concepts/functions/introduction#path-segments.

**Reply by mfix22** [reply]
Thanks @samsisle — I am not sure if that is a broken link or not, but I don't understand how to implement your suggestion from that link.

**Reply by mfix22** [reply]
Found it! https://vercel.com/docs/concepts/functions/serverless-functions#path-segments

**Reply by mfix22** [reply]
@samsisle is there a way to create optional path segments? I have the path segment working [here](https://github.com/carbon-app/carbon/pull/1262), but I want the index (`/api/image`) to also work, but instead I get a 404. Creating a `api/image/index.js` file would probably work, but that will create a whole new lambda, which seems incorrect.

**Reply by mfix22** [reply]
I have found lots of posts like this (https://github.com/vercel/next.js/issues/22342), but there is no clear answer when using it w/ Vercel. 

My PR here https://github.com/carbon-app/carbon/pull/1277 shows that it is not working.

**Reply by samcx** [reply]
Unfortunately, we do not have optional path segments for `/api` Serverless Functions, which isn't the case for `pages/api` (https://nextjs.org/docs/api-routes/dynamic-api-routes) Serverless Functions.

You could try using a rewrite:

```
"rewrites": [{ "source": "/api/image", "destination": "/api/image/index" }]
```

I'm not exactly sure if this would be a viable option for you though (not sure if this destination path is needed?).

**Reply by mfix22** [reply]
Thanks @samsisle! Will you post here if either: 
1. You enable running puppeteer on Next.js API routes
2. I am able to use optional path segments on `/api` Serverless functions

?

**Reply by samcx** [reply]
1. If there’s not enough space to use within ‘pages/api’, you can try from ‘/api’. Generally, we don’t recommend puppeteer on a Serverless Function.
2. We do not have optional path segments for ‘/api’ Serverless Functions.

**Reply by mfix22** [reply]
Sorry, I understand 🙂 I was just wondering if you would update this channel if either of those things change.

**Reply by samcx** [reply]
Oh my bad 😓 — totally didn't understand what you meant there! Will definitely update you if this changes in the future.
